### PR TITLE
Added support for Kinesis

### PR DESCRIPTION
**Pull request by autometa101**
Adding API Support for Kinesis.
It will support all APIs from https://docs.aws.amazon.com/kinesis/latest/APIReference/API_Operations.html

### CODE REVIEW

File: tests/internal/kinesis.js
Comment (oleiade) [inline tests/internal/kinesis.js:85]: Was this console.log statement left on purpose ❓

File: README.md
Comment (oleiade) [inline README.md:264]:
```suggestion
Consult the `KinesisClient` [dedicated k6 documentation page](https://k6.io/docs/javascript-api/jslib/aws/kinesisclient) for more details on its methods and how to use it.
```

I would prefer to follow the same sentence structure as the other examples. We will add the documentation page once this is released 👍🏻 🙇🏻

**Review by oleiade — CHANGES_REQUESTED** [review]
Great work 🎉 👏🏻 Thanks a lot for your contribution, much appreciated 🙇🏻 

I've left some comments regarding how the implementation has been approached. Nothing major, but I would still like to address it before we merge it. Let me know if you have any questions, if I can support you in any ways 👍🏻

File: src/internal/kinesis.ts
Comment (autometa101) [reply src/internal/kinesis.ts:61]: Thanks @oleiade , I have started with individual functions, then realise this would mean to either replicate entire SDK functions or have limited functionality. so, utilising documentation of AWS made sense to me.

on casing standard, I can play gimmick to make it similar, I just kept it as standard as I can see in the page.
https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Operations.html
I can update based based on review opinion and repo standard.

File: tests/internal/kinesis.js
Comment (autometa101) [reply tests/internal/kinesis.js:103]: GetRecord will work based on shard iterators, this log, was to signify MillisBehindLatest will not be 0 on some of the shard iterators and we may not get data in the same.

I did not "Expect" any thing as part of this function so kept it as log.

File: tests/internal/kinesis.js
Comment (autometa101) [reply tests/internal/kinesis.js:85]: I will rectify this.

File: src/internal/kinesis.ts
Comment (oleiade) [reply src/internal/kinesis.ts:61]: Hey @autometa101 

I appreciate your understanding. We've been working with the AWS APIs long enough now to be able to judge that there is a low consistency over there. It's understandable, considering the sheer amount of APIs they have. As a result, we've set our own rules and guidelines, mostly aligned with the rest of the k6 ecosystem.

To be able to merge the Pull Request in `main`, we will indeed require the following:
- That every service method is exposed on the client as a dedicated function (which can call a single shared function under the hood), ideally with some level of documentation (in general, we copy and adapt the one from each action type in the AWS documentation). As I mentioned before, we do that for clarity and user-friendliness but also because it makes our jobs as maintainers easier when it comes to documenting the library on k6's documentation website, for instance, 👍🏻 
- That every method name starts with a lowercase letter 🛩️ 

I imagine this could feel frustrating, but as the library grows, this is how we make sure that contributing to it and using it stays accessible 🙇🏻

File: tests/internal/kinesis.js
Comment (oleiade) [reply tests/internal/kinesis.js:103]: I'd prefer if we could drop it then🤝so that we keep the log output to the point 👍🏻

File: src/internal/kinesis.ts
Comment (autometa101) [reply src/internal/kinesis.ts]: completed, moved to bottom.

File: src/internal/kinesis.ts
Comment (autometa101) [reply src/internal/kinesis.ts:62]: completed, created method for each essential function.

File: src/internal/kinesis.ts
Comment (autometa101) [reply src/internal/kinesis.ts]: done

File: tests/internal/kinesis.js
Comment (autometa101) [reply tests/internal/kinesis.js]: removed.

File: tests/internal/kinesis.js
Comment (autometa101) [reply tests/internal/kinesis.js]: removed

File: README.md
Comment (autometa101) [reply README.md]: done

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:150]:
```suggestion
     * Creates a Kinesis data stream. 
     *
     * A stream captures and transports data records continuously emitted from different data sources or producers.
     *
```

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:159]:
```suggestion
     * Deletes a Kinesis data stream and all its shards and data.
     *
     * You must shut down any applications operating on the stream before you delete the stream.
     * If an application attempts to operate on a deleted stream, it receives the exception ResourceNotFoundException.
     *
```

Comment (oleiade) [reply src/internal/kinesis.ts:159]: I'd also like for us to support the `ResourceNotFoundException` explicitly in this function, as described by the AWS documentation. You can find examples of how it's been done in other clients such as `S3Client` or `SecretsManager`. I believe you'll need to define a dedicated error kind, and adjust the `_handle_error` function body to explicitly handle the `ResourceNotFoundException` 👍🏻

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:167]:
```suggestion
     * Lists your Kinesis data streams.
     *
```

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:177]:
```suggestion
     * Writes multiple data records into a Kinesis data stream in a single call.
     *
     * Use this operation to send data into the stream for data ingestion and processing.
     *
```

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:187]:
```suggestion
     * Gets data records from a Kinesis data stream's shard.
     *
```

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:197]:
```suggestion
     * Lists the shards in a stream and provides information about each shard.
     *
```

File: src/internal/kinesis.ts
Comment (oleiade) [inline src/internal/kinesis.ts:208]:
```suggestion
     * Gets an Amazon Kinesis shard iterator. A shard iterator expires 5 minutes after it is returned to the requester.
     *
```

Comment (oleiade) [reply src/internal/kinesis.ts:208]: I'm not familiar with Kinesis: this returns an "iterator", which seems to be some id by the looks of it. Do we have any methods of making any use of it, though? Otherwise, we'd drop this method altogether. Once again I'm not familiar with Kinesis, so, looking forward to your 💡 here 🙏🏻

**Review by oleiade — CHANGES_REQUESTED** [review]
Great work with the last batch of changes. Thanks a lot for putting the work into it 🙇🏻 

Here's another batch of comments and ideas I'd like to see addressed. That could come across as a lot, but as maintainers, we'll have to own, document, and maintain this code long-term, and I want to ensure we get everything right.  I hope you understand 🙏🏻 

In general, most of my comments are either questions, cosmetic change requests, or questions. Let me know if you have any questions or feedback 👀 

### A more general comment:

I haven't decided on it yet, but I find the whole `*Request`, `*Response` interfaces a bit much to process, document, and maintain.

Furthermore, it's pretty low-level, and relatively close to the actual underlying request/responses format, whereas, at k6, we deeply care about delivering the most friendly-user interfaces possible. In the context of the jslib-aws, we prefer to provide fewer but higher-level abstractions.

I will provide comments directly on the code by the beginning of next week. But my shallow and overarching comment, as of right now, is that I would prefer to have some types/classes that align more with the data types described by the [AWS Kinesis Data Types Documentation](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_Types.html), and less with the underlying bodies of the requests and response. I would like to see them returned directly by the methods exposed to the user rather than the current `*Response` type. That means that instead of a `ListShardsResponse`, for instance, I'd prefer a method that returns an `Array<Shard>`.

I have weaker opinions on this, but the same goes for the `*Request` types. I see their value and convenience to us, the developers. Still, they also obfuscate the actual API and don't align with our goal for the project to make understanding and using the APIs straightforward. In many cases, those interfaces also contain just a couple of members. They could easily be replaced by single arguments (which are also much easier to document when I update the document website).

I hope these comments make sense. Let me know if you need any clarification or support with it, I would be happy to help 🙇🏻

File: src/internal/kinesis.ts
Comment (autometa101) [reply src/internal/kinesis.ts:62]: I would not like to limit the ability of this library, until all functions are available. 
I hope this fine for functionality as concrete functions are already implemented.

Comment (oleiade) [reply src/internal/kinesis.ts:62]: I understand where you come from @autometa101. However, as mentioned in a previous iteration, we would like to avoid having "magic" code (as we need to document the available API to users in k6's documentation), and we would also like to avoid code meant for a future use case.

Let's, please, drop that call for now 🙇🏻

**Comment by oleiade** [comment]
Hey @autometa101, thanks for the further work you've put into this 🤝 and sorry for coming back to this so late, we had a company event in the meantime. We truly value your contribution and have been thinking about how to approach merging it in the main branch in the best conditions. The conclusion was that we would like to merge this PR soonish and incorporate the changes and documentation we think are necessary ourselves. The feature will therefore land in the next upcoming version of the library which has no release date set just yet 👍🏻 


Let me know if you have any opposition to this plan, questions, or feedback. Otherwise, we'll just proceed. Cheers! 🚀

**Comment by autometa101** [comment]
Hi @oleiade,

I was extensively using this library to publish few million messages in last days, 
where I found 2 issues as following.

1. Too big request error, where k6/http post is throwing up the request error. 
2. AWS is throwing error EOF, for the similar reason.

However, I was not able to continue working on the requested changes to remove all methods and allow only limited with details. If there is no changes in direction, I am planning to continue development in next week.
[p.s As User, I found very useful to just add the direct call to API if method is not available.]

**Pull request closed without merging**
